### PR TITLE
S3Target via Smart Open

### PR DESCRIPTION
**Pull request by tokicnikolaus**
Implement S3 target using smart_open library that will upload files to the S3 in chunks under the hood.

### CODE REVIEW

**Comment by tokicnikolaus** [comment]
Hi @siddhantgoel, 
I used this S3target in my project, which might be useful for others also. 

Used smart_open library since it will do the upload in chunks. Boto3 doesn't have the option to do the direct stream, at least I couldn't find it.

Let me know what you think.

File: streaming_form_data/targets.py [hunk 1]
```diff
@@ -164,3 +165,33 @@ def on_data_received(self, chunk: bytes):
...
+    S3Target implemented using smart_open library. Chunked upload to the S3 target.
+    """
+
+    def __init__(self, file_path, mode, transport_params=None, **kwargs):
+        super().__init__(**kwargs)
+        self._file_path = file_path
+        self._mode = mode
+        self._transport_params = transport_params
+        self.data_received_size = 0
+        self._f_out = None
```
Comment (siddhantgoel) [inline streaming_form_data/targets.py:181]: Let's rename this to `self._fd` to keep things consistent with the other targets (eg. `FileTarget`).

File: streaming_form_data/targets.py [hunk 2]
```diff
@@ -164,3 +165,33 @@ def on_data_received(self, chunk: bytes):
...
+    """
+    S3Target implemented using smart_open library. Chunked upload to the S3 target.
+    """
+
+    def __init__(self, file_path, mode, transport_params=None, **kwargs):
+        super().__init__(**kwargs)
+        self._file_path = file_path
+        self._mode = mode
+        self._transport_params = transport_params
+        self.data_received_size = 0
```
Comment (siddhantgoel) [inline streaming_form_data/targets.py:180]: We're not really using this variable. I wonder if we can get rid of it?

File: tests/test_targets.py [hunk 3]
```diff
@@ -255,3 +260,43 @@ def test_custom_target_not_sent():
 
     assert target.value == b''
     assert target.multipart_filename is None
+
+
+class TestS3TargetUpload(TestCase):
```
Comment (siddhantgoel) [inline tests/test_targets.py:265]: Let's use `pytest` style tests written as plain functions. The `boto3` client and `mock_s3` could be added as fixtures that the test function makes use of.

```python
def test_s3_upload(mock_s3, mock_client):
  ...
```

**Review by siddhantgoel — CHANGES_REQUESTED** [review]
Awesome, TIL about `smart_open`! I like how simple it makes implementing such an S3 target.

I've left some minor comments on the target class and would love to have the test written using the pytest style. Otherwise I'm looking forward to merging and releasing this soon!

**Comment by tokicnikolaus** [comment]
@siddhantgoel Thanks for the feedback, I changed it based on your comments, please check  it again and let me know if anything else is needed!

**Comment by tokicnikolaus** [comment]
@siddhantgoel I think all linters are a bit happier now. Since I need to wait for you to allow the workflow to run, let me know if something more pops up!

**Comment by siddhantgoel** [comment]
@tokicnikolaus thanks for the updates, but `black` is somehow still not happy. If you run `make lint` locally it should run all the linters and show you what's failing.

I'm also not sure why Github asks me to approve and run the workflow every time, but I have a bit more time on my hands at the moment so I should be able to click the button rather quickly now!

**Comment by siddhantgoel** [comment]
Thanks! It seems like now pytest is complaining about a missing dependency. You can run `make pip-compile` to update `requirements.txt` after which it should go through.

**Review by siddhantgoel — APPROVED** [review]

**Comment by siddhantgoel** [comment]
@tokicnikolaus awesome, looks like everything's green now! I'll merge this and try to publish a release in the next days.

**Comment by siddhantgoel** [comment]
Just published v1.12.0 which includes this feature. Thanks for your work on this!

**Pull request merged**